sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,352 @@
+                /*
+                 * If the active partition ownership map is empty, this is the first time an event processor is
+                 * running or all Event Processors are down for this Event Hub, consumer group combination. All
+                 * partitions in this Event Hub are available to claim. Choose a random partition to claim ownership.
+                 */
+                claimOwnership(partitionOwnershipMap, partitionIds.get(RANDOM.nextInt(numberOfPartitions)));
+                return;
+            }
+
+            /*
+             * Create a map of owner id and a list of partitions it owns
+             */
+            Map<String, List<PartitionOwnership>> ownerPartitionMap = activePartitionOwnershipMap.values()
+                .stream()
+                .collect(
+                    Collectors.groupingBy(PartitionOwnership::ownerId, mapping(Function.identity(), toList())));
+
+            // add the current event processor to the map if it doesn't exist
+            ownerPartitionMap.putIfAbsent(this.ownerId, new ArrayList<>());
+
+            /*
+             * Find the minimum number of partitions every event processor should own when the load is
+             * evenly distributed.
+             */
+            int numberOfActiveEventProcessors = ownerPartitionMap.size();
+            logger.info("Number of active event processors {}", ownerPartitionMap.size());
+
+            int minPartitionsPerEventProcessor = numberOfPartitions / numberOfActiveEventProcessors;
+
+            /*
+             * If the number of partitions in Event Hub is not evenly divisible by number of active event processors,
+             * a few Event Processors may own 1 additional partition than the minimum when the load is balanced. Calculate
+             * the number of event processors that can own additional partition.
+             */
+            int numberOfEventProcessorsWithAdditionalPartition = numberOfPartitions % numberOfActiveEventProcessors;
+
+            logger.info("Expected min partitions per event processor = {}, expected number of event "
+                + "processors with additional partition = {}", minPartitionsPerEventProcessor, numberOfEventProcessorsWithAdditionalPartition);
+
+            if (isLoadBalanced(minPartitionsPerEventProcessor, numberOfEventProcessorsWithAdditionalPartition,
+                ownerPartitionMap)) {
+                // If the partitions are evenly distributed among all active event processors, no change required.
+                logger.info("Load is balanced");
+                return;
+            }
+
+            if (!shouldOwnMorePartitions(minPartitionsPerEventProcessor, ownerPartitionMap)) {
+                // This event processor already has enough partitions and shouldn't own more.
+                logger.info("This event processor owns {} partitions and shouldn't own more",
+                    ownerPartitionMap.get(ownerId));
```

Comment:
Do you mean to print the number of partitions owned, or the array? Right now I think it will do the array and not the length.